spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+# MF2.0 compromise syntax
+
+# Intro
+
+This syntax builds on the one from https://github.com/unicode-org/message-format-wg/pull/230
+but modified to address
+[@markusicu’s comments there](https://github.com/unicode-org/message-format-wg/pull/230#issuecomment-1116903103).
+
+# Basic syntax
+
+Messages need to delineate between literal text, placeholders, and other “code”.
+We should start in “code mode” and always enclose “patterns” (text+placeholders) in curly braces.
+```
+{Hello world!}
+{Hello {$name}!}
+```
+
+This is unusual for formatting syntaxes, but useful.
+We anyway need to support selecting from among multiple patterns,
+and delimiting the patterns makes it unambiguous
+what white space is part of the pattern vs. serves as delimiters of “code” tokens.
+For consistency, we should always enclose a pattern,
+even if the message consists only of that pattern.
+That also helps with embedding messages in various resource file formats,
+because they can freely trim surrounding white space without
+requiring escapes when a message pattern wants to start or end with spaces.
+
+By contrast, consider the experience with the existing ICU MessageFormat syntax
+which does start in “text mode”.
+ICU MessageFormat has pioneered the selection among multiple patterns based on run-time arguments.
+It represents selection using complex placeholders,
+which has the side effect of allowing literal text and other placeholders
+before and after the top-level selection placeholder.
+However, for reliable translations,
+there should be no translatable contents before or after the selection placeholder;
+instead, each selectable pattern should form one complete “translation unit”.
+Because the existing ICU MessageFormat starts in “text mode”,
+even though it looks like there is no extraneous text,
+spurious white space creeps in from developers’ line breaking of long message strings.
+The remedy is to always use syntax to indicate the start of translatable contents.
+
+We use curly braces to delimit patterns because
+`{}` are the paired ASCII punctuation characters least commonly used in normal text.
+For the same reason, we also use them for embedding placeholders in patterns.
+
+Literal text can use any characters except for curly braces,
+and except for the backslash, which we use as usual for escaping.
+That is, the only special characters inside a pattern are `{}\`.
+The only allowed escape sequences are `\{`, `\}`, and `\\`.
+It is an error if `\` is followed by any other character.
+
+The message syntax does not use `'` or `"`,
+so that it is easy to hard-code message strings in programming language source code.
```

Comment:
How is it intended to scale to preservation of rich syntax in source code? 
I think the paradigm that imperative source code that is used to localize UI and can hard-code textual strings is misrepresentative for how UI architecture is evolving.

1) The main UI language is not Java or C++, it's some declarative markup. If you want to think of how to design MF2.0 syntax to make it easy to hard-code, we need to look at declarative UI languages (APL, HTML, JSX, declarative SwiftUI, Flutter etc.), not imperative programming languages.
2) The content is not just a string. It can be markup as well.

Instead of optimizing for:

```js
let value = mf2.format("Hello World");
label.textContent = value;
```

the hard-coded message in source will be more like:

```html
<button l10n-id="key1">Hello World</button>
```

or even:

```html
<p l10n-id="key1">
  Click <a href="https://www.google.com">here <img src="..."/></a> to go to the next level.
</p>
```

When tools extract from such markup source string into MF2 resource, they can replace `<a>` with `{html:a}` etc. but it will be a pattern with composition of strings and markup elements, not a single string embedded as a string literal in Java or C++.

Reply:
> Presumably messages can use these characters internally (in literals), since apostrophe can't always be avoided in real text and since double-quotes are sometimes desired. For languages that consider these characters to be part of syntax, the escaping is on a higher level, right?

Yes. If it's hardcoded in, say, a Java string, then you use Java escapes to embed double quotes.

Of course, it's not good practice to hardcode messages in source code files, but people do, at least for prototyping.

It's also not good practice to use ASCII apostrophes and ASCII double quotes in user-facing messages :-) Now that we are no longer limited to ISO 8859-x charsets, we have been pushing developers and translators to use typographic’ “quotes”.

Reply:
> 1. The main UI language is not Java or C++, it's some declarative markup. If you want to think of how to design MF2.0 syntax to make it easy to hard-code, we need to look at declarative UI languages (APL, HTML, JSX, declarative SwiftUI, Flutter etc.), not imperative programming languages.

We can look at multiple source formats, but we can't avoid all possible syntax characters.
People do hardcode strings in C++/Java/Python etc. as well as in XML and other markup.
I am ok avoiding both quotes and `<>` angle brackets for MF2 syntax. I left plenty of TODO's for such details.

> When tools extract from such markup source string into MF2 resource, they can replace `<a>` with `{html:a}` etc. but it will be a pattern with composition of strings and markup elements, not a single string embedded as a string literal in Java or C++.

The input to the message formatting function, and the input and output of a localization system, is *a string* with syntax. The contents of the message is part of the string, one way or another.

Reply:
> The input to the message formatting function, and the input and output of a localization system, is a string with syntax. (...) The contents of the message is part of the string.

One of the changes we are introducing in MF2.0 I am advocating for is the paradigm shift away from this mental model.
The input/output is not a string. It is a sequence of parts, some of which are strings, some of which are markup.

Reply:
> > The input to the message formatting function, and the input and output of a localization system, is a string with syntax. (...) The contents of the message is part of the string.
> 
> One of the changes we are introducing in MF2.0 I am advocating for is the paradigm shift away from this mental model. The input/output is not a string. It is a sequence of parts, some of which are strings, some of which are markup.

The output can be a string, or a format-to-parts structure of some kind. (As discussed in March.)

The input has to be a string as far as the syntax is concerned, *or else it makes no sense to even talk about “the syntax”*. This string syntax is also what would show up in CLDR data files if they used MF2.0, and as the input to formatting libraries that take a format string as an argument -- which developers will expect to be able to do.

There can of course be other types of inputs = representations of the data model, such as an XLIFF spec etc. They necessarily have different syntaxes.

Reply:
> The input has to be a string as far as the syntax is concerned

Correct, but that string, when extracted from a source code, will be modified when placed in a CLDR data file or MF2 data file.

So when extracting from HTML `<p l10n-id="foo">Hello </img></p>` will convert to MF2 `Hello {html:img}`. Same can happen to quotes, so "easieness" of hardcoding in source code does not depend on use of quote or double-quotes in MF2 syntax.

Reply:
> > The input has to be a string as far as the syntax is concerned
> 
> Correct, but that string, when extracted from a source code, will be modified when placed in a CLDR data file or MF2 data file. [...] Same can happen to quotes, so "easieness" of hardcoding in source code does not depend on use of quote or double-quotes in MF2 syntax.

Sure, but it makes sense to design the syntax so that common cases are easy, or don't need any transformations/escaping.

Reply:
> Sure, but it makes sense to design the syntax so that common cases are easy, or don't need any transformations/escaping.

Agree. The question that I'm not sure if we're aligned on is whether common cases will be imperative code with string literal encoding MF2.0 message in source code, or some templating UI with MF2.0 message in source code.
It very well may be that we'll want to follow your logic for both (so, avoid '"', "'", and "<" as common in embedded source code syntaxes?)

Reply:
I think we should be prepared for both cases. We as well as choosing characters that are infrequent in messages, we try to miss most of the characters that would have special meanings inside of programming language strings, and the same for templating languages. 

One unpleasant feature of use \{, \}, \\ for escaping is that inside of some programming language's strings we'd have to use \\\{, \\\}, and \\\\. I don't see a good way around that (ideas welcome), but on the other handle most such programming language users are probably used to that. 

Minimizing the 'special' characters, keeping them down to {, }, and \ would help there also.